lib/internal/bootstrap/loaders.js

```diff
@@ -208,6 +242,93 @@
         NativeModule.require;
       fn(this.exports, requireFn, this, process);
 
+      if (config.experimentalModules) {
+        this.exportKeys = Object.keys(this.exports);
+
+        const update = (property, value) => {
+          if (this.reflect !== undefined && this.exportKeys.includes(property))
+            this.reflect.exports[property].set(value);
+        };
+
+        const methodWrapMap = new WeakMap();
+        const methodUnwrapMap = new WeakMap();
+
+        const wrap = (target, name, value) => {
+          if (typeof value !== 'function' || !isNative(value)) {
+            return value;
+          }
+
+          if (methodWrapMap.has(value))
+            return methodWrapMap.get(value);
+
+          const p = new Proxy(value, {
+            apply: (t, thisArg, args) => {
+              if (thisArg === proxy || (this.reflect !== undefined &&
+                  this.reflect.namespace !== undefined &&
+                  thisArg === this.reflect.namespace)) {
+                thisArg = target;
+              }
+              return ReflectApply(t, thisArg, args);
+            },
+            __proto__: null,
+          });
+
+          methodWrapMap.set(value, p);
+          methodUnwrapMap.set(p, value);
+
+          return p;
+        };
+
+        const proxy = new Proxy(this.exports, {
+          set: (target, prop, value, receiver) => {
+            if (typeof value === 'function')
+              value = methodUnwrapMap.get(value) || value;
+            if (ReflectSet(target, prop, value, receiver)) {
+              update(prop, ReflectGet(target, prop, receiver));
```

Comment:
not quite sure i understand this one, but i do know that `Object.create(module.exports).asdf = 'asdf'` doesn't change anything on `module.exports`

Reply:
@guybedford
> Exactly, so it shouldn't call the update function, whereas right now it does I believe.

I cannot reproduce your concern.
```
import { isArray } from "util"
> Object.create(util).isArray  = 'asdf'
'asdf'
> isArray
[Function: isArray]
> util.isArray
[Function: isArray]
```

Reply:
@bmeck I cannot reproduce an issue with the implementation. 
```js
import f from "./foo.js"
// f is module.exports = {}

f.toString = null
console.log(f.toString) // null
console.log(Object.prototype.toString) // [Function]
```

Can you provide a named import usage example so that a test can be written.

Reply:
so to fix this, i should do like
```js
if (target === receiver)
  update(prop, ReflectGet(target, prop, receiver));

Reply:
@devsnek Let's wait for a repro before tweaking please.

Reply:
```mjs
import fs,{readFile} from 'fs';
Object.create(fs).readFile = 123;
```

Calls `update()` but doesn't produce a affect right now. The code always is reading off the `target` which seems to avoid the bug in question for data properties, but seems like it would cause issues if it had an accessor that used the receiver. I'm not seeing any accessors right now on a glance but it is doing an extra call that seems odd.

Reply:
It would be a bummer to add superstitious bits to the implementation without a real repo case.
Is there a way we can create a test for accessors (like a dummy built-in) to include so we know for sure if there's an issue and test for regressions in the future?

Reply:
@jdalton it isn't really superstitious and you can see the odd extraneous update call using the inspector. To my knowledge no internal builtins exist purely for testing purposes of the nature you describe.

Reply:
@bmeck Checking more this seem ok too
```js
import f from "util"

var count = 0

var p = {
  get foo() { return ++count },
  set foo(v) {
    Reflect.defineProperty(p, 'foo', {
      configurable: true,
      enumerable: true,
      writable: true,
      value: v
    })
  }
}

f.foo = 'blah'
console.log(count) // 0
console.log(f.foo) // "blah"
console.log(p.foo) // 1
console.log(p.foo) // 2
p.foo = "done"
console.log(p.foo) // "done"
```

If the `module.exports` had an accessor then it would be an own property and be part of the `reflect.exports` and so should be accessed. I don't think there's a bug here. My concern is that by placing guards for things that aren't real issues we could potentially block something intended to pass.

Reply:
Would adding a test like the one be enough for you @bmeck.

Reply:
Your example has p as the receiver, but my concern is accessors on the target. If we had a test to ensure that targets are safe, that would be fine. I would still prefer we not call unnecessary `update()`s though.

Reply:
@bmeck So a test like this?
```js
import events from 'events'
import { defaultMaxListeners } from 'events'

var count = 0

Reflect.defineProperty(Function.prototype, 'defaultMaxListeners', {
  configurable: true,
  enumerable: true,
  get: function() { return ++count },
  set: function(v) {
    Reflect.defineProperty(this, 'foo', {
      configurable: true,
      enumerable: true,
      writable: true,
      value: v
    })
  }
})

console.log(events.defaultMaxListeners) // 10
console.log(++events.defaultMaxListeners) // 11
console.log(++events.defaultMaxListeners) // 12
console.log(defaultMaxListeners) // 12
console.log(Function.prototype.defaultMaxListeners) // 1
```

Reply:
@jdalton that doesn't have the Proxy as the target with questionable accessors, you need something that gets wrapped as if it were a builtin module that looks like:

```mjs
var bar;
const worstTargetEver = {
  set foo(value) {
    Object.defineProperty(this, 'foo', {value}) // not on target, need to ensure it isn't picked up
    delete worstTargetEver.foo; // target no longer has a descriptor for 'foo' after this
    bar = value; // need to update() bar
  };
  get bar() { return bar; }
}
const testProxy = new Proxy(, builtinModuleTraps)
```

This has a few things going on:

1. we need to add a test to ensure we don't get `value` since that is an own property on the receiver so we can prevent regressions.
2. right now it could update to `Object.prototype.foo` since there isn't a descriptor anymore but we still do a non-guarded Get after the Set. I didn't see any existing things in core that would be problematic. We could document that such behaviors should never be parts of core APIs instead of adding guards though. I don't think this behavior is lintable so it would be some tribal knowledge to pick up :-/.
3. this somewhat shows a problem if setters are defining other properties that might need to be updated. I'm not sure we can guard against this except with some documentation that such types of setters should never land in core. Reading core yesterday I didn't find existing code that would be problematic, but we should document that such things should never land in core due to this proxy behavior making things go out of sync. Should probably document that with how live getter properties shouldn't land in core.

I'm not too happy about tribal knowledge and the need to avoid doing specific APIs, but wouldn't block if some tests show number 1 isn't regressing and we add documentation about 2/3 somewhere since I don't think we can statically test that or enforce it with the test harness in a feasible manner.

Reply:
Looks like userland can do some stuff since we allow `defineProperty` trapping that exhibits these behaviors, I guess any sort of documentation about these limits would be public and not apply just to core:

```mjs
// show-out-of-sync.mjs
import fs, * as _ from 'fs';
Object.prototype.readFile = 123;
Object.defineProperty(fs, 'readFile', {
  set(v) {delete fs.readFile},
  configurable: true
});
fs.readFile = null;
console.log('fs', fs.readFile, 'from namespace', _.readFile);
Object.prototype.readFile = 456;
console.log('fs', fs.readFile, 'from namespace', _.readFile);
```

Reply:
> Looks like userland can do some stuff since we allow defineProperty trapping that exhibits these behaviors,

In your example above you have removed the own property of `fs` when you assigned it `fs.readFile=null` because in the setter it performed `delete fs.readFile`. So when you logged both `fs.readFile` values they had a `fs.hasOwnProperty("readFile")` of `false`. The `deleteProperty` trap should have updated the ns values to be `undefined` though so we should check that out.

Reply:
In my own implementation I get the expected results:
```
fs 123 from namespace undefined
fs 456 from namespace undefined
```

Reply:
It's all good news! In my local tests for the `worstTargetEver` things are looking good. I'll help add tests.

> 1. we need to add a test to ensure we don't get value since that is an own property on the receiver so we can prevent regressions.

This look good

```js
import { foo, bar } from "worstTargetEver"
import w from "worstTargetEver"

w.foo = 3
foo // undefined
bar // 3
```

> 2. right now it could update to Object.prototype.foo since there isn't a descriptor anymore but we still do a non-guarded Get after the Set. I didn't see any existing things in core that would be problematic.

The example above did not result in a value getting assigned to `Object.prototype.foo`
```
 Reflect.has(Object.prototype, "foo") // false
```

> 3. this somewhat shows a problem if setters are defining other properties that might need to be updated.

The example above works as expected, to me at least.

I'll work with @devsnek to add tests and ensure the PR results align with my local.

Reply:
@bmeck I had the same concern at the beginning, but because of the limited scope of this PR (only applies to core modules) I didn't think we have to be that watertight, because we don't really have those weird setters/getters.

Reply:
@bmeck @TimothyGu As mentioned above ([here](https://github.com/nodejs/node/pull/20403#discussion_r185500673) and [here](https://github.com/nodejs/node/pull/20403#discussion_r185506543)) there should be no synchronization mismatch or unexpected gotchas once I get a chance to update this PR and add tests.